vllm/v1/spec_decode/medusa.py

```diff
@@ -42,11 +42,18 @@ def propose(
         blocks = self.model(target_hidden_states)
         logits = self.model.compute_logits(blocks)
 
-        # Get draft tokens and transpose the result
-        # TODO(woosuk): OPTIMIZATION: Return GPU tensor without GPU-CPU
-        # synchronization.
-        draft_tokens = [logit.argmax(dim=-1).tolist() for logit in logits]
-        return [list(row) for row in zip(*draft_tokens)]
+        # Non-primary DP/TP ranks emit an empty placeholder.
+        if not logits:
+            return []
+
+        # Argmax per head keeps memory light and cuts GPU→CPU syncs to one pass.
+        argmax_tokens = [
+            logit.argmax(dim=-1) for logit in logits
+        ]  # List of [batch_size] tensors
+        draft_tokens_gpu = torch.stack(argmax_tokens, dim=1)  # [batch_size, num_heads]
```

Comment:
Is `logits` a list or tuple coming into this? I'm curious if it would be possible to stack first and then compute a batched argmax? Especially if it's already a stacked tensor coming in, it will be trivial. Otherwise it's a tradeoff worth benchmarking